src/Util/ExAuth.php

```diff
@@ -1,306 +1,355 @@
+				$uid = User::authenticate($aUser, $aCommand[3]);
+				$Error = $uid === false;
+			} else {
+				$this->writeLog(LOG_WARNING, 'user not found: ' . $sUser);
+				$Error = true;
+				$uid = -1;
+			}
+			if ($Error) {
+				$this->writeLog(LOG_INFO, 'check against alternate password for ' . $sUser . '@' . $aCommand[2]);
+				$sPassword = PConfig::get($uid, 'xmpp', 'password', null, true);
+				$Error = ($aCommand[3] != $sPassword);
+			}
+		} else {
+			$Error = true;
+		}
+
+		// If the hostnames doesn't match or there is some failure, we try to check remotely
+		if ($Error) {
+			$Error = !$this->checkCredentials($aCommand[2], $aCommand[1], $aCommand[3], true);
+		}
+
+		if ($Error) {
+			$this->writeLog(LOG_WARNING, 'authentification failed for user ' . $sUser . '@' . $aCommand[2]);
+			fwrite(STDOUT, pack('nn', 2, 0));
+		} else {
+			$this->writeLog(LOG_NOTICE, 'authentificated user ' . $sUser . '@' . $aCommand[2]);
+			fwrite(STDOUT, pack('nn', 2, 1));
+		}
+	}
+
+	/**
+	 * @brief Check remote credentials via HTTP(S)
+	 *
+	 * @param string $host The hostname
+	 * @param string $user Username
+	 * @param string $password Password
+	 * @param boolean $ssl Should the check be done via SSL?
+	 *
+	 * @return boolean Are the credentials okay?
+	 */
+	private function checkCredentials($host, $user, $password, $ssl)
+	{
+		$this->writeLog(LOG_INFO, 'external credential check for ' . $user . '@' . $host);
+
+		$url = ($ssl ? 'https' : 'http') . '://' . $host . '/api/account/verify_credentials.json?skip_status=true';
+
+		$ch = curl_init();
+		curl_setopt($ch, CURLOPT_URL, $url);
+		curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+		curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 5);
+		curl_setopt($ch, CURLOPT_HEADER, true);
+		curl_setopt($ch, CURLOPT_NOBODY, true);
+		curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+		curl_setopt($ch, CURLOPT_USERPWD, $user . ':' . $password);
+
+		curl_exec($ch);
+		$curl_info = @curl_getinfo($ch);
+		$http_code = $curl_info['http_code'];
+		curl_close($ch);
+
+		$this->writeLog(LOG_INFO, 'external auth for ' . $user . '@' . $host . ' returned ' . $http_code);
+
+		return $http_code == 200;
+	}
+
+	/**
+	 * @brief Set the hostname for this process
+	 *
+	 * @param string $host The hostname
+	 */
+	private function setHost($host)
+	{
+		if (!empty($this->host)) {
+			return;
+		}
+
+		$this->writeLog(LOG_INFO, 'Hostname for process ' . getmypid() . ' is ' . $host);
+
+		$this->host = $host;
+
+		$lockpath = Config::get('jabber', 'lockpath');
+		if (is_null($lockpath)) {
+			return;
+		}
+
+		$this->pidfile = new Pidfile($lockpath, $host);
+		if ($this->pidfile->isRunning()) {
+			$oldpid = $this->pidfile->pid();
+			$this->writeLog(LOG_INFO, 'Process ' . $oldpid . ' was running for ' . $this->pidfile->runningTime() . ' seconds and will now be killed');
+			$this->pidfile->kill();
+
+			// Wait until the other process is hopefully killed
+			sleep(2);
+
+			$this->pidfile = new Pidfile($lockpath, $host);
```

Comment:
Can you please add a `refresh()` or `reload()` method to `Pidfile` that perform the file same check as in the constructor so that you don't have to create a new instance?

Reply:
Okay, I did some changes.